R/teal_slice.R

```diff
@@ -483,44 +530,51 @@ c.teal_slices <- function(...) {
   )
 }
 
+# store method for `teal_slices` object
+#' @param x `teal_slices` object
+#' @param file a `character(1)` with the storage destination name
+#' @export
+#' @rdname teal_slice
+#' @keywords internal
+#'
+store.teal_slices <- function(x, file = NULL, ...) {
+  checkmate::assert_character(file, len = 1, null.ok = TRUE)
+
+  cat(format(x, show_all = TRUE, pretty = FALSE), file = file)
+}
 
-# format method for teal_slices
+# format method for `teal_slices`
 #' @export
 #' @rdname teal_slice
 #' @keywords internal
 #'
-format.teal_slices <- function(x, show_all = FALSE, ...) {
-  res <- character(0)
-  for (i in seq_along(x)) {
-    ind <- names(x)[i]
-    if (is.null(ind)) ind <- sprintf("[[%d]]", i)
-    res <- append(res, ind)
-    res <- append(res, format(x[[i]], show_all = show_all, ...))
-  }
+format.teal_slices <- function(x, show_all = FALSE, pretty = TRUE, ...) {
 
-  includes <- attr(x, "include_varnames")
-  if (length(includes) > 0L) {
-    res <- append(res, "\nfilterable variables:")
-    for (i in seq_along(includes)) {
-      res <- append(res, sprintf(" $ %s: %s", names(includes)[i], toString(includes[[i]])))
-    }
-  }
+  x_format <- lapply(x, format, show_all = show_all, pretty = pretty)
 
-  excludes <- attr(x, "exclude_varnames")
-  if (length(excludes) > 0L) {
-    res <- append(res, "\nnon-filterable variables:")
-    for (i in seq_along(excludes)) {
-      res <- append(res, sprintf(" $ %s: %s", names(excludes)[i], toString(excludes[[i]])))
-    }
-  }
+  attributes <- extract_attrs(x, show_all = show_all)
 
-  ct <- attr(x, "count_type")
-  res <- append(res, sprintf("\ncount type: %s", ct))
-  paste(c(res, ""), collapse = "\n")
+  paste0(c(unlist(x_format), attributes), collapse = '\n')
+
+}
+
+# attributes extraction for `teal_slices` object
+#' @param x `teal_slice` object
+#' @param show_all should NULL attributes be shown
+#' @keywords internal
+#'
+extract_attrs <- function(x, show_all = FALSE){
+  attributes <- list(
+    include_varnames = attr(x, "include_varnames"),
+    exclude_varnames = attr(x, "exclude_varnames"),
+          count_type = attr(x, "count_type")
+  )
+  if(!show_all) attributes <- Filter(Negate(is.null), attributes)
+  paste0('attributes\n', to_json(attributes))
```

Comment:
Just use `attributes(x)` and drop "class". Will also work if new attributes are added to `teal_slices`.

Reply:
amazing. was experimenting with `attributes` but I did figure out that I was running this function on the improper object : D and that's why I didn't notice it can help. great catch